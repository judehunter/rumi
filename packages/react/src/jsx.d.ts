import {createRumi} from '@rumi/core/src/createRumi';

type CSS = Parameters<ReturnType<typeof createRumi>['css']>[0];

declare module 'react' {
  /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
  interface DOMAttributes<T> {
    css?: CSS;
  }
  interface HTMLAttributes<T> extends AriaAttributes, DOMAttributes<T> {
    css?: CSS;
  }
}

declare global {
  namespace JSX {
    interface IntrinsicAttributes {
      css?: CSS;
    }
  }
}

export {};
