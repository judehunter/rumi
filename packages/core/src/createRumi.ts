import {stringify} from '@stitches/stringify';
import {toHash} from './hash';
import {all as mergeAll} from 'deepmerge';

type CSSBaseStyles = {
  color: string;
  backgroundColor: string;
  display: 'block' | 'inline' | 'flex';
  justifyContent: any;
};

type UtilFunction = (value: any) => Partial<CSSBaseStyles>;

type RumiTheme = any;

type Media = Record<string, string>;

type RumiConfig = {
  utils: Record<string, UtilFunction>;
  theme?: RumiTheme;
  media?: Media;
  prefix?: string;
};

type ID<T> = {[Prop in keyof T]: T[Prop]};

export const createRumi = <T extends RumiConfig>(cfg: T) => {
  type StylesObject = CSSBaseStyles & {
    [key: string]: string | Partial<StylesObject>;
  } & {
    [Prop in keyof T['utils']]: Parameters<T['utils'][Prop]>[0];
  } & {
    [Prop in keyof T['media']]: Partial<StylesObject>;
  }; // remove partial here and add deep partial elsewhere

  type StylesPrimitive =
    | null
    | undefined
    | boolean
    | number
    | string
    | Partial<StylesObject>;

  type TOrArrayT<T> = T | T[];

  let stylesheet = {};

  const getCssText = ({reset = true} = {}) => {
    const ret = Object.values(stylesheet).join('');
    if (reset) stylesheet = {};
    return ret;
  };

  const addToStylesheet = (hash: string, cssString: string) => {
    stylesheet[hash] = cssString;

    // do different things depending on the environment
    if (globalThis.document) {
      let styleTag = globalThis.document.head.querySelector('style[data-rumi]');
      if (styleTag == null) {
        styleTag = globalThis.document.createElement('style');
        styleTag.setAttribute('data-rumi', '');

        globalThis.document.head.appendChild(styleTag);
      }

      styleTag.appendChild(globalThis.document.createTextNode(cssString));
    }
  };

  const transformSpecialProperties = (styles: any) => {
    let newObj = {};

    for (const key of Object.keys(styles)) {
      // console.log('key', key);

      const foundUtilFunction = Object.entries(cfg.utils).find(
        ([k]) => k === key,
      )?.[1];

      if (foundUtilFunction) {
        newObj = {...newObj, ...foundUtilFunction(styles[key])};
        continue;
      }

      let foundAtRule = Object.entries(cfg.media ?? {}).find(
        ([k]) => k === key,
      )?.[1];

      if (foundAtRule) {
        foundAtRule = `@media ${foundAtRule}`;
        newObj[foundAtRule] = styles[key];
        if (newObj[foundAtRule] instanceof Object) {
          newObj[foundAtRule] = transformSpecialProperties(newObj[foundAtRule]);
        }
        continue;
      }

      if (styles[key] instanceof Object) {
        newObj[key] = transformSpecialProperties(styles[key]);

        continue;
      }

      newObj[key] = styles[key];
    }

    return newObj;
  };

  const mergeStyles = (styles: any[]) => {
    return mergeAll(styles);
  };

  const css = (styles: TOrArrayT<StylesPrimitive>) => {
    const enabledStyles = (Array.isArray(styles) ? styles : [styles]).filter(
      (x) => x && x instanceof Object,
    );

    const transformedStyles = enabledStyles.map((x) =>
      transformSpecialProperties(x),
    );

    const mergedStyles = mergeStyles(transformedStyles);

    const className = `.${cfg.prefix ? cfg.prefix + '-' : ''}rumi-${toHash(
      mergedStyles,
    )}`;

    if (!Object.keys(stylesheet).includes(className)) {
      addToStylesheet(
        className,
        stringify({
          [className]: mergedStyles,
        }),
      );
    }

    const classNameGetter = () => {
      return className;
    };
    return classNameGetter;
  };

  return {css, getCssText};
};
