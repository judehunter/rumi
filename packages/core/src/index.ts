export {createRumi} from './createRumi';
export {flushCache} from './stylesheet';
