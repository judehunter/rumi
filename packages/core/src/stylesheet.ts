let styleTag: HTMLStyleElement | null = null;

let stylesheet: Record<string, {cssString: string; styles: any}> = {};

export let cache: Record<string, string> = {};

export const flushCache = () => (cache = {});

export const addToStylesheetAndCache = (
  className: string,
  cssString: string,
  styles?: any,
) => {
  // if this class is not in the stylesheet yet
  if (!Object.keys(stylesheet).includes(className)) {
    stylesheet[className] = {cssString, styles};

    if (globalThis.document) {
      if (styleTag === null) {
        styleTag = globalThis.document.createElement('style');
        styleTag.setAttribute('data-rumi', '');

        globalThis.document.head.appendChild(styleTag);
      }

      styleTag.appendChild(globalThis.document.createTextNode(cssString));
    }
  }
};

export const getStylesheetCSSText = ({reset = true} = {}) => {
  const ret = Object.values(stylesheet)
    .map((x) => x.cssString)
    .join('');
  if (reset) stylesheet = {};
  return ret;
};
