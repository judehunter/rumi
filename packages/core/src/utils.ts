export type TOrArrayT<T> = T | T[];
