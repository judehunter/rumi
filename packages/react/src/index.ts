import {createRumi} from '@rumi/core';

console.log(createRumi);
