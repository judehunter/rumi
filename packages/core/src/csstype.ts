import {StandardProperties, SimplePseudos} from 'csstype';
import {RumiConfig} from './createRumi';
import {TOrArrayT} from './utils';
// import * as Native from './css';

// export {Native};

export type CSSBaseStyles = StandardProperties<
  string | number,
  string | number
>;
// export interface CSSBaseStyles
//   extends Native.StandardLonghandProperties,
//     Native.StandardShorthandProperties,
//     Native.SvgProperties {}

export type CSSProperties = {
  [K in keyof CSSBaseStyles]:
    | CSSBaseStyles[K]
    | Array<Extract<CSSBaseStyles[K], string>>;
};

export type CSSPseudos<T extends RumiConfig> = {
  [K in SimplePseudos as `&${K}`]?: TOrArrayT<StylesObject<T>>;
};

export type CSSNested<T extends RumiConfig> = {
  [K in
    | `& *`
    | '& > *'
    | '& + *'
    | `& ~ *`
    | `${string}&${string}`
    | `${string}@${string}`]?: TOrArrayT<StylesObject<T>>;
};

export type CSSUtils<T extends RumiConfig> = {
  [Prop in keyof T['utils']]?: T['utils'][Prop] extends (...args: any) => any
    ? Parameters<T['utils'][Prop]>[0]
    : never;
};

export type CSSMedia<T extends RumiConfig> = {
  [Prop in keyof T['media']]?: StylesObject<T>;
};
export interface StylesObjectInterface<T extends RumiConfig>
  extends CSSProperties,
    CSSPseudos<T>,
    CSSNested<T> {}

export type StylesObject<T extends RumiConfig> =
  | (StylesObjectInterface<T> & CSSUtils<T>)
  | {styles: StylesObject<T>};

// export type StylesObject<T extends RumiConfig> =
//   | ({[Prop in keyof CSSBaseStyles]?: CSSBaseStyles[Prop]} & {
//       [Prop in keyof T['media']]?: StylesObject<T>;
// } & {
//   [Prop in keyof T['utils']]?: T['utils'][Prop] extends (
//     ...args: any
//   ) => any
//     ? Parameters<T['utils'][Prop]>[0]
//     : never;
// } & {
//       [K in Contains<'&'>]?: StylesObject<T>;
//     } & {
//       [K in keyof CSSNestedSelectors]?: StylesObject<T>;
//     })
//   | {styles: StylesObject<T>};
