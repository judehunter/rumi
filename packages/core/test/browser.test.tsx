/**
 * @jest-environment jsdom
 */

import {createRumi} from '../src/createRumi';

// describe('browser', () => {
//   test('simple styles', () => {
//     const {css} = createRumi({
//       utils: {
//         bg: (val) => ({backgroundColor: val}),
//       },
//     });
//     css({
//       bg: 'red',
//     });
//     css({
//       bg: 'violet',
//     });

//     expect(document.head.innerHTML).toMatchSnapshot();
//   });
// });
