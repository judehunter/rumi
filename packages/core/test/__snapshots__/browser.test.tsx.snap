// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`browser simple styles 1`] = `"<style data-rumi=\\"\\">.rumi-dkEfkH{background-color:red;}.rumi-jKSHCh{background-color:violet;}</style>"`;
